File: backend/app/api/v1/documents.py
from fastapi import APIRouter, Depends, File, Form, Query, Request, UploadFile
from loguru import logger
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.config import (
    RATE_LIMIT_DEFAULT,
    RATE_LIMIT_MODERATE,
    RATE_LIMIT_SEVERE,
    RATE_LIMIT_VERY_STRICT,
)
from app.core.cache import cache_delete_pattern, cache_get, cache_set
from app.core.exceptions import ConflictError
from app.core.middleware import limiter
from app.database import get_db
from app.db.user import User
from app.schemas.common import ok, paginated_ok
from app.schemas.document import DocumentOut
from app.services import document_service

router = APIRouter(prefix="/documents", tags=["documents"])


@router.post("/upload")
@limiter.limit(RATE_LIMIT_VERY_STRICT)
async def upload_document(
    request: Request,
    file: UploadFile = File(...),
    kb_id: int = Form(...),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """上传文档到知识库。业务逻辑下沉到 document_service.upload_document。"""
    doc, task = await document_service.upload_document(file, kb_id, user, db)
    logger.info(f"Document uploaded: id={doc.id} kb={kb_id} user={user.id} name={doc.filename}")
    # 缓存失效：清除该用户的文档列表缓存
    await cache_delete_pattern(f"doc:list:{user.id}:*")
    return ok(
        data={
            "document_id": doc.id,
            "status": "pending",
            "task_id": task.id,
        }
    )


@router.get("")
@limiter.limit(RATE_LIMIT_DEFAULT)
async def list_documents(
    request: Request,
    kb_id: int | None = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """分页查询文档列表。使用 Redis 缓存，TTL 5 分钟。"""
    # 尝试命中缓存
    cache_key = f"doc:list:{user.id}:{kb_id}:{page}:{page_size}"
    cached = await cache_get(cache_key)
    if cached is not None:
        return paginated_ok(
            items=cached["items"], total=cached["total"], page=page, page_size=page_size
        )

    docs, total = await document_service.list_documents(user.id, db, kb_id, page, page_size)
    items = [DocumentOut.model_validate(d).model_dump() for d in docs]
    # 写入缓存
    await cache_set(cache_key, {"items": items, "total": total})
    return paginated_ok(items=items, total=total, page=page, page_size=page_size)


@router.get("/{doc_id}")
@limiter.limit(RATE_LIMIT_DEFAULT)
async def get_document(
    doc_id: int,
    request: Request,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    doc = await document_service.get_document(doc_id, user.id, db)
    return ok(data=DocumentOut.model_validate(doc).model_dump())


@router.get("/{doc_id}/progress")
@limiter.limit(RATE_LIMIT_DEFAULT)
async def get_progress(
    doc_id: int,
    request: Request,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """获取文档处理进度。业务逻辑（Redis 缓存读写、进度计算）下沉到
    document_service.get_progress。本层仅做参数绑定与响应格式化。"""
    data = await document_service.get_progress(doc_id, user.id, db)
    return ok(data=data)


@router.delete("/{doc_id}")
@limiter.limit(RATE_LIMIT_DEFAULT)
async def delete_document(
    doc_id: int,
    request: Request,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # Check if document is currently being processed
    doc = await document_service.get_document_for_write(doc_id, user.id, db)
    if doc.status in ("parsing", "chunking", "embedding"):
        raise ConflictError(message="文档正在处理中，无法删除，请稍候")
    await document_service.delete_document(doc_id, user.id, db)
    logger.info(f"Document deleted: id={doc_id} user={user.id}")
    # 缓存失效
    await cache_delete_pattern(f"doc:list:{user.id}:*")
    return ok(message="Deleted")


@router.post("/{doc_id}/reparse")
@limiter.limit(RATE_LIMIT_SEVERE)
async def reparse_document(
    doc_id: int,
    request: Request,
    force: bool = Query(False, description="是否强制重新解析（跳过乐观锁）"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    # 使用 service 的原子锁, 防止并发重复触发
    doc, task = await document_service.reparse_document(doc_id, user.id, db, force=force)
    logger.info(f"Document reparse: id={doc_id} user={user.id} force={force}")
    # 缓存失效
    await cache_delete_pattern(f"doc:list:{user.id}:*")
    return ok(data={"document_id": doc.id, "task_id": task.id})


@router.get("/{doc_id}/preview")
@limiter.limit(RATE_LIMIT_MODERATE)
async def preview_document(
    doc_id: int,
    request: Request,
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Preview document content with pagination (lines).

    业务逻辑（文件解析、分页计算、错误处理）下沉到
    document_service.preview_document。本层仅做参数绑定与响应格式化。
    """
    data = await document_service.preview_document(doc_id, user.id, db, page, page_size)
    return ok(data=data)

File: frontend/src/api/documents.ts
import client, { extractData } from './client';
import { getWithOptionalSignal } from './helpers';
import { globalT } from '../i18n';
import type { Document, DocumentProgress } from '../types';

export interface DocumentPreviewData {
  filename: string;
  file_type: string;
  content: string;
  page: number;
  page_size: number;
  total_lines: number;
  total_pages: number;
}

export const documentApi = {
  /** 获取文档列表 (kbId 省略时后端返回所有有权限的文档, 实现真正的服务端分页) */
  async list(kbId?: number, page = 1, pageSize = 20, signal?: AbortSignal): Promise<{ items: Document[]; total: number; page: number; page_size: number; total_pages: number }> {
    return getWithOptionalSignal<{ items: Document[]; total: number; page: number; page_size: number; total_pages: number }>(
      '/documents',
      {
        ...(kbId !== undefined ? { kb_id: kbId } : {}),
        page,
        page_size: pageSize,
      },
      signal,
    );
  },

  /** 获取文档详情 */
  async get(docId: number): Promise<Document> {
    const res = await client.get(`/documents/${docId}`);
    return extractData<Document>(res);
  },

  /** 上传文档 */
  async upload(
    kbId: number,
    file: File,
    onProgress?: (loaded: number, total: number) => void,
    signal?: AbortSignal,
  ): Promise<{ document_id: number; status: string; task_id: string }> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('kb_id', String(kbId));

    try {
      const res = await client.post(
        '/documents/upload',
        formData,
        {
          timeout: 120000, // 上传超时 2 分钟
          onUploadProgress: (progressEvent) => {
            if (onProgress && progressEvent.total) {
              onProgress(progressEvent.loaded, progressEvent.total);
            }
          },
          signal,
        },
      );
      return extractData<{ document_id: number; status: string; task_id: string }>(res);
    } catch (error) {
      // 主动 abort 不应显示上传失败提示 (用户已主动关闭 Modal)
      if (signal?.aborted) throw error;
      // client 拦截器已统一提取服务器 message；上传场景无 message 时使用 i18n 兜底
      if (error instanceof Error && error.message) {
        throw error;
      }
      throw new Error(globalT('document.uploadFailed'));
    }
  },

  /** 删除文档 */
  async delete(docId: number): Promise<void> {
    await client.delete(`/documents/${docId}`);
  },

  /** 重新解析文档 */
  async reparse(docId: number, force: boolean = false): Promise<{ document_id: number; task_id: string }> {
    const res = await client.post(`/documents/${docId}/reparse`, null, {
      params: { force },
    });
    return extractData(res);
  },

  /** 获取文档处理进度 */
  async getProgress(docId: number, signal?: AbortSignal): Promise<DocumentProgress> {
    return getWithOptionalSignal<DocumentProgress>(`/documents/${docId}/progress`, undefined, signal);
  },

  /** 预览文档内容 */
  async preview(docId: number, page = 1, pageSize = 50, signal?: AbortSignal): Promise<DocumentPreviewData> {
    const res = await client.get(`/documents/${docId}/preview`, {
      params: { page, page_size: pageSize },
      signal,
    });
    return extractData<DocumentPreviewData>(res);
  },
};

export default documentApi;
